File: apps/api/src/telegram/approvals.ts
import type { Signal } from "@prisma/client";
import { resolveRiskConfig } from "../config/resolve";
import { openLiveTrade } from "../execution/liveTrade";
import { openPaperTrade } from "../execution/paperTrading";
import { prisma } from "../lib/prisma";
import { publishEvent } from "../lib/realtime";
import { calculatePositionSize } from "../risk/riskEngine";

function isLiveBroker(): boolean {
  return (process.env.BROKER ?? "paper").trim().toLowerCase() === "exness";
}
import {
  defaultChatId,
  editMessageText,
  isConfigured,
  sendMessage,
} from "./telegram";

/** Escape the few characters Telegram's HTML parse_mode treats specially. */
function esc(s: string): string {
  return s.replace(/&/g, "&amp;").replace(/</g, "&lt;").replace(/>/g, "&gt;");
}

function num(v: { toString(): string } | null | undefined): number {
  return v == null ? 0 : Number(v.toString());
}

function fmt(symbol: string, v: number): string {
  const dp = symbol === "EURUSD" ? 4 : symbol === "BTCUSD" ? 0 : 2;
  return v.toLocaleString("en-US", { minimumFractionDigits: dp, maximumFractionDigits: dp });
}

function account() {
  return { accountBalance: Number(process.env.PAPER_ACCOUNT_BALANCE ?? "10000") };
}

/** Build the full alert text (plan + AI reasoning) from data the gate produced. */
export async function formatAlert(signal: Signal, ttlMin: number): Promise<string> {
  const cfg = await resolveRiskConfig(signal.strategyName, signal.symbol);
  const { accountBalance } = account();
  const entry = num(signal.entryPrice);
  const stop = num(signal.stopLoss);
  const target = num(signal.takeProfit);
  const risk = Math.abs(entry - stop);
  const reward = Math.abs(target - entry);
  const rr = risk > 0 ? reward / risk : 0;

  let sizeLine = "—";
  try {
    const sized = calculatePositionSize(accountBalance, cfg.riskPerTradePct, entry, stop);
    sizeLine =
      `${sized.lotSize.toFixed(4)} units  (risk $${sized.riskAmount.toFixed(2)} = ` +
      `${cfg.riskPerTradePct}% of $${accountBalance.toLocaleString()})`;
  } catch {
    /* leave dash */
  }

  const dot = signal.direction === "LONG" ? "🟢" : "🔴";
  const reasoning = (signal.aiReasoning || "").trim().slice(0, 600);

  return [
    `${dot} <b>SIGNAL</b> — ${esc(signal.symbol)} ${esc(signal.timeframe)} · ${signal.direction} · ${esc(signal.strategyName ?? "—")}`,
    `Mode: CONFIRM · expires in ${ttlMin}m`,
    ``,
    `<b>PLAN</b>`,
    `• Entry   ${fmt(signal.symbol, entry)}`,
    `• Stop    ${fmt(signal.symbol, stop)}`,
    `• Target  ${fmt(signal.symbol, target)}`,
    `• R:R     1:${rr.toFixed(2)}`,
    `• Size    ${sizeLine}`,
    ``,
    `<b>WHY</b> (AI score ${signal.confidenceScore}/100)`,
    esc(reasoning) || "—",
  ].join("\n");
}

export interface RequestApprovalResult {
  created: boolean;
  reason?: string;
}

/**
 * Create an Approval(PENDING) for a CONFIRM-mode signal and send the Telegram
 * alert with Approve/Reject buttons. Fail-safe: if Telegram isn't configured or
 * the send fails, the Approval is still recorded so the signal is auditable and
 * the expiry sweep can clean it up; the signal never auto-opens as a fallback.
 */
export async function requestApproval(signal: Signal, ttlMin: number): Promise<RequestApprovalResult> {
  const existing = await prisma.approval.findUnique({ where: { signalId: signal.id } });
  if (existing) return { created: false, reason: "approval_exists" };

  const chatId = defaultChatId() ?? "";
  const expiresAt = new Date(Date.now() + ttlMin * 60_000);

  const approval = await prisma.approval.create({
    data: { signalId: signal.id, chatId, expiresAt },
  });

  if (!isConfigured()) {
    console.warn(`[approvals] Telegram not configured — approval ${approval.id} created without alert`);
    return { created: true, reason: "telegram_not_configured" };
  }

  const text = await formatAlert(signal, ttlMin);
  const messageId = await sendMessage(chatId, text, [
    [
      { text: "✅ Approve", callback_data: `apv:${approval.id}` },
      { text: "❌ Reject", callback_data: `rej:${approval.id}` },
    ],
  ]);

  if (messageId) {
    await prisma.approval.update({ where: { id: approval.id }, data: { messageId } });
  }
  return { created: true };
}

export interface DecisionResult {
  ok: boolean;
  outcome: "approved" | "rejected" | "already_decided" | "expired" | "not_found" | "open_failed";
  message: string;
}

/**
 * Apply an Approve/Reject decision coming back from the Telegram webhook.
 * Idempotent: a second tap on an already-decided/expired approval is a no-op.
 */
export async function applyApprovalDecision(
  approvalId: string,
  approve: boolean,
  decidedBy: string,
): Promise<DecisionResult> {
  const approval = await prisma.approval.findUnique({
    where: { id: approvalId },
    include: { signal: true },
  });
  if (!approval) return { ok: false, outcome: "not_found", message: "Approval not found." };

  if (approval.status !== "PENDING") {
    return { ok: false, outcome: "already_decided", message: `Already ${approval.status.toLowerCase()}.` };
  }
  if (approval.expiresAt.getTime() < Date.now()) {
    await prisma.$transaction([
      prisma.approval.update({ where: { id: approval.id }, data: { status: "EXPIRED" } }),
      prisma.signal.update({ where: { id: approval.signalId }, data: { status: "CANCELLED" } }),
    ]);
    return { ok: false, outcome: "expired", message: "This signal already expired." };
  }

  const stamp = new Date();
  if (!approve) {
    await prisma.$transaction([
      prisma.approval.update({
        where: { id: approval.id },
        data: { status: "REJECTED", decidedBy, decidedAt: stamp },
      }),
      prisma.signal.update({ where: { id: approval.signalId }, data: { status: "CANCELLED" } }),
    ]);
    void publishEvent({ type: "signal", symbol: approval.signal.symbol });
    return { ok: true, outcome: "rejected", message: `❌ Rejected by ${decidedBy}` };
  }

  // Approve → the authoritative risk re-size happens inside the trade opener.
  const opened = isLiveBroker()
    ? await openLiveTrade(approval.signalId)
    : await openPaperTrade(approval.signalId);
  if (opened.status !== "opened") {
    return { ok: false, outcome: "open_failed", message: `Could not open: ${opened.reason ?? "unknown"}` };
  }
  await prisma.approval.update({
    where: { id: approval.id },
    data: { status: "APPROVED", decidedBy, decidedAt: stamp },
  });
  void publishEvent({ type: "trade", symbol: approval.signal.symbol });
  return { ok: true, outcome: "approved", message: `✅ Approved by ${decidedBy} · trade opened` };
}

export interface ExpirySummary {
  expired: number;
}

/**
 * Expire any PENDING approval past its TTL: mark EXPIRED, cancel the signal, and
 * stamp the Telegram message. Run every minute from the scheduler.
 */
export async function expireStaleApprovals(): Promise<ExpirySummary> {
  const stale = await prisma.approval.findMany({
    where: { status: "PENDING", expiresAt: { lt: new Date() } },
    include: { signal: { select: { symbol: true } } },
    take: 100,
  });

  let expired = 0;
  for (const a of stale) {
    await prisma.$transaction([
      prisma.approval.update({ where: { id: a.id }, data: { status: "EXPIRED" } }),
      prisma.signal.update({ where: { id: a.signalId }, data: { status: "CANCELLED" } }),
    ]);
    if (a.messageId) {
      await editMessageText(a.chatId, a.messageId, "⌛ <b>Expired</b> — not taken.");
    }
    void publishEvent({ type: "signal", symbol: a.signal.symbol });
    expired += 1;
  }
  return { expired };
}

File: services/backend/app/integrations/telegram/approvals.py
"""Human-in-the-loop approvals — port of `apps/api/src/telegram/approvals.ts`.

A CONFIRM-mode signal gets an `Approval(PENDING)` row plus a Telegram card with
Approve/Reject buttons. Fail-safe throughout: if Telegram isn't configured or
the send fails, the Approval is still recorded so the signal is auditable and the
expiry sweep can clean it up — the signal never auto-opens as a fallback.
"""
from __future__ import annotations

import logging
from datetime import timedelta
from typing import Any

from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from ...core.ids import new_id
from ...core.realtime import publish_event
from ...core.serialization import num
from ...core.settings import get_settings
from ...db.enums import ApprovalStatus, SignalStatus
from ...db.models import Approval, Signal
from ...domain.config.resolve import resolve_risk_config
from ...domain.risk.engine import calculate_position_size
from ...jobs.clock import as_aware_utc, naive_utcnow, utcnow
from .client import (
    default_chat_id,
    edit_message_text,
    esc,
    is_configured,
    send_message,
)

log = logging.getLogger("backend.approvals")


def _fmt(symbol: str, value: float) -> str:
    """Price formatting per instrument, mirroring the TS `toLocaleString` call."""
    dp = 4 if symbol == "EURUSD" else 0 if symbol == "BTCUSD" else 2
    return f"{value:,.{dp}f}"


async def format_alert(session: AsyncSession, signal: Signal, ttl_min: int) -> str:
    """Build the full alert text (plan + AI reasoning) from data the gate produced."""
    cfg = await resolve_risk_config(session, signal.strategyName, signal.symbol)
    balance = get_settings().paper_account_balance
    entry = num(signal.entryPrice)
    stop = num(signal.stopLoss)
    target = num(signal.takeProfit)
    risk = abs(entry - stop)
    reward = abs(target - entry)
    rr = reward / risk if risk > 0 else 0.0

    size_line = "—"
    try:
        sized = calculate_position_size(balance, cfg.riskPerTradePct, entry, stop)
        pct = cfg.riskPerTradePct
        pct_text = str(int(pct)) if float(pct).is_integer() else str(pct)
        size_line = (
            f"{sized.lotSize:.4f} units  (risk ${sized.riskAmount:.2f} = "
            f"{pct_text}% of ${balance:,.0f})"
        )
    except ValueError:
        pass  # leave the dash

    dot = "🟢" if signal.direction.value == "LONG" else "🔴"
    reasoning = (signal.aiReasoning or "").strip()[:600]

    return "\n".join(
        [
            f"{dot} <b>SIGNAL</b> — {esc(signal.symbol)} {esc(signal.timeframe)} · "
            f"{signal.direction.value} · {esc(signal.strategyName or '—')}",
            f"Mode: CONFIRM · expires in {ttl_min}m",
            "",
            "<b>PLAN</b>",
            f"• Entry   {_fmt(signal.symbol, entry)}",
            f"• Stop    {_fmt(signal.symbol, stop)}",
            f"• Target  {_fmt(signal.symbol, target)}",
            f"• R:R     1:{rr:.2f}",
            f"• Size    {size_line}",
            "",
            f"<b>WHY</b> (AI score {signal.confidenceScore}/100)",
            esc(reasoning) or "—",
        ]
    )


async def request_approval(
    session: AsyncSession, signal: Signal, ttl_min: int
) -> dict[str, Any]:
    """Create an `Approval(PENDING)` and send the Telegram alert with buttons."""
    if get_settings().api_shadow_mode:
        return {"created": False, "reason": "shadow_mode_no_telegram"}

    existing = (
        await session.execute(select(Approval).where(Approval.signalId == signal.id))
    ).scalar_one_or_none()
    if existing is not None:
        return {"created": False, "reason": "approval_exists"}

    chat_id = default_chat_id() or ""
    approval = Approval(
        id=new_id(),
        signalId=signal.id,
        status=ApprovalStatus.PENDING,
        chatId=chat_id,
        expiresAt=naive_utcnow() + timedelta(minutes=ttl_min),
        createdAt=naive_utcnow(),
    )
    session.add(approval)
    await session.commit()

    if not is_configured():
        log.warning(
            "[approvals] Telegram not configured — approval %s created without alert",
            approval.id,
        )
        return {"created": True, "reason": "telegram_not_configured"}

    text = await format_alert(session, signal, ttl_min)
    message_id = await send_message(
        chat_id,
        text,
        [
            [
                {"text": "✅ Approve", "callback_data": f"apv:{approval.id}"},
                {"text": "❌ Reject", "callback_data": f"rej:{approval.id}"},
            ]
        ],
    )
    if message_id:
        approval.messageId = message_id
        await session.commit()
    return {"created": True}


async def apply_approval_decision(
    session: AsyncSession, approval_id: str, approve: bool, decided_by: str
) -> dict[str, Any]:
    """Apply an Approve/Reject decision from the Telegram webhook.

    Idempotent: a second tap on an already-decided/expired approval is a no-op.
    """
    from ...domain.execution.live_trade import open_live_trade
    from ...domain.execution.paper_trading import open_paper_trade

    approval = (
        await session.execute(select(Approval).where(Approval.id == approval_id))
    ).scalar_one_or_none()
    if approval is None:
        return {"ok": False, "outcome": "not_found", "message": "Approval not found."}

    if approval.status != ApprovalStatus.PENDING:
        return {
            "ok": False,
            "outcome": "already_decided",
            "message": f"Already {approval.status.value.lower()}.",
        }
    if as_aware_utc(approval.expiresAt) < utcnow():
        approval.status = ApprovalStatus.EXPIRED
        approval.signal.status = SignalStatus.CANCELLED
        await session.commit()
        return {"ok": False, "outcome": "expired", "message": "This signal already expired."}

    stamp = naive_utcnow()
    if not approve:
        approval.status = ApprovalStatus.REJECTED
        approval.decidedBy = decided_by
        approval.decidedAt = stamp
        approval.signal.status = SignalStatus.CANCELLED
        await session.commit()
        await publish_event("signal", symbol=approval.signal.symbol)
        return {"ok": True, "outcome": "rejected", "message": f"❌ Rejected by {decided_by}"}

    # Approve → the authoritative risk re-size happens inside the trade opener.
    opener = open_live_trade if get_settings().is_live_broker else open_paper_trade
    opened = await opener(session, approval.signalId)
    if opened.status != "opened":
        return {
            "ok": False,
            "outcome": "open_failed",
            "message": f"Could not open: {opened.reason or 'unknown'}",
        }
    approval.status = ApprovalStatus.APPROVED
    approval.decidedBy = decided_by
    approval.decidedAt = stamp
    await session.commit()
    await publish_event("trade", symbol=approval.signal.symbol)
    return {
        "ok": True,
        "outcome": "approved",
        "message": f"✅ Approved by {decided_by} · trade opened",
    }


async def expire_stale_approvals(session: AsyncSession) -> dict[str, int]:
    """Expire any PENDING approval past its TTL: mark EXPIRED, cancel the signal,
    and stamp the Telegram message. Run every minute from the scheduler."""
    stale = (
        (
            await session.execute(
                select(Approval)
                .where(
                    Approval.status == ApprovalStatus.PENDING,
                    Approval.expiresAt < naive_utcnow(),
                )
                .limit(100)
            )
        )
        .scalars()
        .all()
    )

    expired = 0
    for approval in stale:
        approval.status = ApprovalStatus.EXPIRED
        approval.signal.status = SignalStatus.CANCELLED
        symbol = approval.signal.symbol
        await session.commit()
        if approval.messageId:
            await edit_message_text(
                approval.chatId, approval.messageId, "⌛ <b>Expired</b> — not taken."
            )
        await publish_event("signal", symbol=symbol)
        expired += 1
    return {"expired": expired}
